### PR TITLE
Centrifuge db build

### LINKED ISSUE #2956: New module: centrifuge/build
### Is there an existing module for this?

- [x] I have searched for the existing module

### Is there an open PR for this?

- [X] I have searched for existing PRs

### Is there an open issue for this?

- [X] I have searched for existing issues

### Are you going to work on this?

- [X] If I'm planning to work on this module, I added myself to the `Assignees` to facilitate tracking who is working on the module

### PR DESCRIPTION
**Pull request by sarah-buddle**
## PR checklist

Closes #2956   

- [x] This comment contains a description of changes (with reason).
- [ ] If you've fixed a bug or added code that should be tested, add tests!
- [x] If you've added a new tool - have you followed the module conventions in the [contribution docs](https://github.com/nf-core/modules/tree/master/.github/CONTRIBUTING.md)
- [ ] If necessary, include test data in your PR.
- [x] Remove all TODO statements.
- [x] Emit the `versions.yml` file.
- [x] Follow the naming conventions.
- [x] Follow the parameters requirements.
- [x] Follow the input/output options guidelines.
- [x] Add a resource `label`
- [x] Use BioConda and BioContainers if possible to fulfil software requirements.
- Ensure that the test works with either Docker / Singularity. Conda CI tests can be quite flaky:
  - [ ] `PROFILE=docker pytest --tag <MODULE> --symlink --keep-workflow-wd --git-aware`
  - [x] `PROFILE=singularity pytest --tag <MODULE> --symlink --keep-workflow-wd --git-aware`
  - [x] `PROFILE=conda pytest --tag <MODULE> --symlink --keep-workflow-wd --git-aware`

### CODE REVIEW

File: modules/nf-core/centrifuge/build/main.nf [hunk 1]
```diff
@@ -0,0 +1,41 @@
+process CENTRIFUGE_BUILD {
+    tag "$meta.id"
+    label 'process_high'
+
+    conda "bioconda::centrifuge=1.0.4"
+    container "${ workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?
+        'https://depot.galaxyproject.org/singularity/centrifuge-core:1.0.4--h5b5514e_0':
+        'quay.io/biocontainers1.0.4--hd03093a_0' }"
```
Comment (sofstam) [inline modules/nf-core/centrifuge/build/main.nf:8]:
```suggestion
        'quay.io/biocontainers/centrifuge:1.0.4--hd03093a_0' }"
```

File: modules/nf-core/centrifuge/build/main.nf [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+
+    input:
+    tuple val(meta), path(fasta)
+    path conversion_table
+    path taxonomy
+    path name_table
+
+    output:
+    tuple val(meta), path("*.cf"), emit: cf
```
Comment (sofstam) [inline modules/nf-core/centrifuge/build/main.nf:17]:
```suggestion
    tuple val(meta), path("*.cf").    , emit: cf
```

File: modules/nf-core/centrifuge/build/main.nf [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+    path "versions.yml"           , emit: versions
+
+    when:
+    task.ext.when == null || task.ext.when
+
+    script:
+    def args = task.ext.args ?: ''
+    def prefix = task.ext.prefix ?: "${meta.id}"
+
```
Comment (sofstam) [inline modules/nf-core/centrifuge/build/main.nf:26]:
```suggestion
```

File: modules/nf-core/centrifuge/build/main.nf [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+    conda "bioconda::centrifuge=1.0.4"
+    container "${ workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?
+        'https://depot.galaxyproject.org/singularity/centrifuge-core:1.0.4--h5b5514e_0':
+        'quay.io/biocontainers1.0.4--hd03093a_0' }"
+
+    input:
+    tuple val(meta), path(fasta)
+    path conversion_table
+    path taxonomy
```
Comment (sofstam) [inline modules/nf-core/centrifuge/build/main.nf:13]:
```suggestion
    path taxonomy_tree
```

Looking at the documentation, I am thinking if we should rename taxonomy like that instead

File: modules/nf-core/centrifuge/build/main.nf [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+    def args = task.ext.args ?: ''
+    def prefix = task.ext.prefix ?: "${meta.id}"
+
+    """
+    centrifuge-build \\
+    -p $task.cpus \\
+    $fasta \\
+    ${prefix} \\
+    --conversion-table $conversion_table \\
+    --taxonomy-tree $taxonomy \\
```
Comment (sofstam) [inline modules/nf-core/centrifuge/build/main.nf:33]:
```suggestion
    --taxonomy-tree $taxonomy_tree \\
```

File: modules/nf-core/centrifuge/build/meta.yml [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+input:
+  - meta:
+      type: map
+      description: |
+        Groovy Map containing sample information
+        e.g. [ id:'test', single_end:false ]
+  
+  - fasta:
+      type: file
+      description: fasta file containing sequences to be used in database
```
Comment (sofstam) [inline modules/nf-core/centrifuge/build/meta.yml:23]:
```suggestion
      description: Fasta file containing sequences to be used in centrifuge database.
```

File: modules/nf-core/centrifuge/build/meta.yml [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+      pattern: "*.{fasta}"
```
Comment (sofstam) [inline modules/nf-core/centrifuge/build/meta.yml:24]:
```suggestion
      pattern: "*.{fasta,fna}"
```

File: modules/nf-core/centrifuge/build/meta.yml [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+
+  - conversion_table:
+      type: file
+      description: a tab-separated sequence ID to taxonomy ID mapping
```
Comment (sofstam) [inline modules/nf-core/centrifuge/build/meta.yml:28]:
```suggestion
      description: A tab-separated file with sequence ID to taxonomy ID mapping
```

File: modules/nf-core/centrifuge/build/meta.yml [hunk 9]
```diff
@@ -0,0 +1,60 @@
...
+      pattern: "*.{map}"
+
+  - taxonomy_tree:
+      type: file
+      description: \t|\t-separated file mapping taxonomy. Typically nodes.dmp from the NCBI taxonomy dump. Links taxonomy IDs to their parents
```
Comment (sofstam) [inline modules/nf-core/centrifuge/build/meta.yml:33]:
```suggestion
      description: A \t|\t-separated file mapping taxonomy. Typically nodes.dmp from the NCBI taxonomy dump. Links taxonomy IDs to their parents
```

File: modules/nf-core/centrifuge/build/meta.yml [hunk 10]
```diff
@@ -0,0 +1,60 @@
...
+      pattern: "*.{dmp}"
+
+  - name_table:
+      type: file
+      description: '|'-separated file mapping taxonomy IDs to a name. Typically names.dmp from the NCBI taxonomy dump. Links taxonomy IDs to their scientific name
```
Comment (sofstam) [inline modules/nf-core/centrifuge/build/meta.yml:38]:
```suggestion
      description: A '|'-separated file mapping taxonomy IDs to a name. Typically names.dmp from the NCBI taxonomy dump. Links taxonomy IDs to their scientific name
```

File: modules/nf-core/centrifuge/build/meta.yml [hunk 11]
```diff
@@ -0,0 +1,60 @@
...
+        e.g. [ id:'test', single_end:false ]
+  
+  - versions:
+      type: file
+      description: File containing software versions
+      pattern: "versions.yml"
+
+  - cf:
+      type: file
+      description: Index files of the database
```
Comment (sofstam) [inline modules/nf-core/centrifuge/build/meta.yml:56]:
```suggestion
      description: Index files of the centrifuge database
```

File: tests/config/pytest_modules.yml [hunk 12]
```diff
@@ -1785,6 +1789,10 @@ kofamscan:
   - modules/nf-core/kofamscan/**
   - tests/modules/nf-core/kofamscan/**
 
+kraken2/build:
+  - modules/nf-core/kraken2/build/**
+  - tests/modules/nf-core/kraken2/build/**
```
Comment (sofstam) [inline tests/config/pytest_modules.yml:1794]:
```suggestion
```

Comment (sofstam) [inline tests/config/pytest_modules.yml:1794]:
```suggestion
```

File: tests/config/pytest_modules.yml [hunk 13]
```diff
@@ -1785,6 +1789,10 @@ kofamscan:
   - modules/nf-core/kofamscan/**
   - tests/modules/nf-core/kofamscan/**
 
+kraken2/build:
+  - modules/nf-core/kraken2/build/**
```
Comment (sofstam) [inline tests/config/pytest_modules.yml:1793]:
```suggestion
```

File: tests/config/pytest_modules.yml [hunk 14]
```diff
@@ -2046,6 +2054,10 @@ metabat2/metabat2:
   - modules/nf-core/metabat2/metabat2/**
   - tests/modules/nf-core/metabat2/metabat2/**
 
+metamix:
+  - modules/nf-core/metamix/**
+  - tests/modules/nf-core/metamix/**
```
Comment (sofstam) [inline tests/config/pytest_modules.yml:2059]:
```suggestion
```

File: tests/config/pytest_modules.yml [hunk 15]
```diff
@@ -2046,6 +2054,10 @@ metabat2/metabat2:
   - modules/nf-core/metabat2/metabat2/**
   - tests/modules/nf-core/metabat2/metabat2/**
 
+metamix:
+  - modules/nf-core/metamix/**
```
Comment (sofstam) [inline tests/config/pytest_modules.yml:2058]:
```suggestion
```

File: tests/config/pytest_modules.yml [hunk 16]
```diff
@@ -2046,6 +2054,10 @@ metabat2/metabat2:
   - modules/nf-core/metabat2/metabat2/**
   - tests/modules/nf-core/metabat2/metabat2/**
 
+metamix:
```
Comment (sofstam) [inline tests/config/pytest_modules.yml:2057]:
```suggestion
```

File: tests/modules/nf-core/centrifuge/build/main.nf [hunk 17]
```diff
@@ -0,0 +1,25 @@
...
+include { CENTRIFUGE_BUILD } from '../../../../../modules/nf-core/centrifuge/build/main.nf'
+
+workflow test_centrifuge_build {
+    
+    input = [
+        [ id:'test', single_end:false ], // meta map
+        file( 'https://raw.githubusercontent.com/nf-core/test-datasets/modules/data/genomics/sarscov2/genome/genome.fasta', checkIfExists: true)
+    ]
+
+   //// params.test_data['sarscov2']['genome']['genome_fasta']
```
Comment (sofstam) [inline tests/modules/nf-core/centrifuge/build/main.nf:14]:
```suggestion

```

**Review by sofstam — CHANGES_REQUESTED** [review]
Nice work! A few suggestions :)

File: tests/config/pytest_modules.yml (hunk 13 above)
Comment (sarah-buddle) [reply tests/config/pytest_modules.yml:1793]: Remove files from kraken test

File: modules/nf-core/centrifuge/build/main.nf (hunk 2 above)
Comment (sarah-buddle) [reply modules/nf-core/centrifuge/build/main.nf:17]: What is this for?

Comment (sofstam) [reply modules/nf-core/centrifuge/build/main.nf:17]: Sorry, the `.` is a typo.

File: modules/nf-core/centrifuge/build/main.nf [hunk 18]
```diff
@@ -0,0 +1,40 @@
...
+        'quay.io/biocontainers/centrifuge:1.0.4--hd03093a_0' }"
+
+    input:
+    tuple val(meta), path(fasta)
+    path conversion_table
+    path taxonomy_tree
+    path name_table
+
+    output:
+    tuple val(meta), path("*.cf").    , emit: cf
```
Comment (sofstam) [inline modules/nf-core/centrifuge/build/main.nf:17]:
```suggestion
    tuple val(meta), path("*.cf")   , emit: cf
```

I removed the unnecessary `.`

File: tests/config/pytest_modules.yml [hunk 19]
```diff
@@ -1785,6 +1789,8 @@ kofamscan:
   - modules/nf-core/kofamscan/**
   - tests/modules/nf-core/kofamscan/**
 
+kraken2/build:
```
Comment (sofstam) [inline tests/config/pytest_modules.yml:1792]:
```suggestion
```

This should be removed as well.

File: tests/config/pytest_modules.yml [hunk 20]
```diff
@@ -2046,6 +2052,7 @@ metabat2/metabat2:
   - modules/nf-core/metabat2/metabat2/**
   - tests/modules/nf-core/metabat2/metabat2/**
 
+
```
Comment (sofstam) [inline tests/config/pytest_modules.yml:2055]:
```suggestion
```

File: tests/modules/nf-core/centrifuge/build/main.nf [hunk 21]
```diff
@@ -0,0 +1,24 @@
...
+include { CENTRIFUGE_BUILD } from '../../../../../modules/nf-core/centrifuge/build/main.nf'
+
+workflow test_centrifuge_build {
+    
+    input = [
+        [ id:'test', single_end:false ], // meta map
+        file( 'https://raw.githubusercontent.com/nf-core/test-datasets/modules/data/genomics/sarscov2/genome/genome.fasta', checkIfExists: true)
+    ]
+
+
```
Comment (sofstam) [inline tests/modules/nf-core/centrifuge/build/main.nf:14]:
```suggestion
```

File: tests/modules/nf-core/centrifuge/build/main.nf [hunk 22]
```diff
@@ -0,0 +1,24 @@
...
+    conversion_table = file(params.test_data['sarscov2']['metagenome']['seqid2taxid_map'], checkIfExists: true)
+
```
Comment (sofstam) [inline tests/modules/nf-core/centrifuge/build/main.nf:16]:
```suggestion
```

File: tests/modules/nf-core/centrifuge/build/main.nf [hunk 23]
```diff
@@ -0,0 +1,24 @@
...
+    taxonomy_tree = file(params.test_data['sarscov2']['metagenome']['nodes_dmp'], checkIfExists: true)
+
```
Comment (sofstam) [inline tests/modules/nf-core/centrifuge/build/main.nf:18]:
```suggestion
```

File: tests/modules/nf-core/centrifuge/build/main.nf [hunk 24]
```diff
@@ -0,0 +1,24 @@
...
+    name_table = file(params.test_data['sarscov2']['metagenome']['names_dmp'], checkIfExists: true)
+
+    
+
```
Comment (sofstam) [inline tests/modules/nf-core/centrifuge/build/main.nf:22]:
```suggestion
```

File: tests/modules/nf-core/centrifuge/build/main.nf [hunk 25]
```diff
@@ -0,0 +1,24 @@
...
+    ]
+
+
+    conversion_table = file(params.test_data['sarscov2']['metagenome']['seqid2taxid_map'], checkIfExists: true)
+
+    taxonomy_tree = file(params.test_data['sarscov2']['metagenome']['nodes_dmp'], checkIfExists: true)
+
+    name_table = file(params.test_data['sarscov2']['metagenome']['names_dmp'], checkIfExists: true)
+
+
```
Comment (sofstam) [inline tests/modules/nf-core/centrifuge/build/main.nf:21]:
```suggestion
```

**Review by sofstam — CHANGES_REQUESTED** [review]
A few minor comments and after the tests pass, we can merge 🎉

File: modules/nf-core/centrifuge/build/main.nf [hunk 26]
```diff
@@ -0,0 +1,56 @@
...
+    container "${ workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?
+        'https://depot.galaxyproject.org/singularity/centrifuge:1.0.4_beta--h9a82719_6':
+        'biocontainers/centrifuge:1.0.4_beta--h9a82719_6' }"
+
+    input:
+    tuple val(meta), path(fasta)
+    path conversion_table
+    path taxonomy_tree
+    path name_table
+    path size_table
```
Comment (sofstam) [inline modules/nf-core/centrifuge/build/main.nf:15]: Is this used somewhere as argument?

File: modules/nf-core/centrifuge/build/tests/main.nf.test [hunk 27]
```diff
@@ -0,0 +1,35 @@
...
+    test("sarscov2 - fasta") {
+
+        when {
+            process {
+                """
+                input[0] = [ [id: 'test'], file(params.test_data['sarscov2']['genome']['genome_fasta'], checkIfExists: true) ]
+                input[1] = file(params.test_data['sarscov2']['metagenome']['seqid2taxid_map'], checkIfExists: true)
+                input[2] = file(params.test_data['sarscov2']['metagenome']['nodes_dmp'], checkIfExists: true)
+                input[3] = file(params.test_data['sarscov2']['metagenome']['names_dmp'], checkIfExists: true)
+                input[4] = []
```
Comment (sofstam) [inline modules/nf-core/centrifuge/build/tests/main.nf.test:21]: Is this supposed to be empty?

**Review by sofstam — COMMENTED** [review]
🚀

File: modules/nf-core/centrifuge/build/tests/main.nf.test (hunk 27 above)
Comment (maxulysse) [reply modules/nf-core/centrifuge/build/tests/main.nf.test:21]: I'm guessing it's optional

File: modules/nf-core/centrifuge/build/meta.yml [hunk 28]
```diff
@@ -0,0 +1,68 @@
...
+      type: file
+      description: A \t|\t-separated file mapping taxonomy. Typically nodes.dmp from the NCBI taxonomy dump. Links taxonomy IDs to their parents
+      pattern: "*.{dmp}"
+
+  - name_table:
+      type: file
+      description: A '|'-separated file mapping taxonomy IDs to a name. Typically names.dmp from the NCBI taxonomy dump. Links taxonomy IDs to their scientific name
+      pattern: "*.{dmp}"
+
+  - size_table:
```
Comment (maxulysse) [inline modules/nf-core/centrifuge/build/meta.yml:44]: shouldn't we say it's optional

File: modules/nf-core/centrifuge/build/tests/nextflow.config [hunk 29]
```diff
@@ -0,0 +1,4 @@
+process {
+
```
Comment (sateeshperi) [inline modules/nf-core/centrifuge/build/tests/nextflow.config:2]: can remove empty file

File: modules/nf-core/centrifuge/build/tests/main.nf.test (hunk 27 above)
Comment (jfy133) [reply modules/nf-core/centrifuge/build/tests/main.nf.test:21]: YEs correct! Updated meta

**Review by sofstam — APPROVED** [review]
🚀

**Comment by jfy133** [comment]
Thanks @sarah-buddle for doing the majority of the legwork on that one! We beat a year since beginning :laughing:

**Pull request merged**